urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/server/AbstractINDIServer.java

```diff
@@ -0,0 +1,812 @@
+   */
+  protected void removeDevice(INDIDevice device) {
+    String[] names = device.getNames();
+
+    devices.remove(device);
+    clients.remove(device);
+
+    notifyClientsDeviceRemoved(names);
+
+    driverDisconnected(device.getDeviceIdentifier(), names);
+  }
+
+  /**
+   * Notifies the listening clients that some particular Devices have been
+   * removed by sending
+   * <code>delProperty</code> messages.
+   *
+   * @param deviceNames The names of the Devices that have been removed.
+   */
+  private void notifyClientsDeviceRemoved(String[] deviceNames) {
+    for (int h = 0 ; h < deviceNames.length ; h++) {
+      String deviceName = deviceNames[h];
+
+      String message = "<delProperty device=\"" + deviceName + "\" />";
+
+      ArrayList<INDIDeviceListener> list = this.getClientsListeningToDevice(deviceName);
+
+      for (int i = 0 ; i < list.size() ; i++) {
+        INDIDeviceListener c = list.get(i);
+
+        c.sendXMLMessage(message);
+      }
+
+      ArrayList<INDIDeviceListener> list2 = this.getClientsListeningToSingleProperties(deviceName);
+
+      for (int i = 0 ; i < list2.size() ; i++) {
+        INDIDeviceListener c = list2.get(i);
+
+        c.sendXMLMessage(message);
+      }
+    }
+  }
+
+  /**
+   * Removes a Client from the List of clients. Called by the clients when the
+   * connection is broken.
+   *
+   * @param client The Client to remove.
+   */
+  protected void removeClient(INDIClient client) {
+    clients.remove(client);
+
+    connectionWithClientBroken(client);
+  }
+
+  /**
+   * Checks if a Class is subclass of INDIDriver.
+   *
+   * @param c The class to check
+   * @return
+   * <code>true</code> if the Class inherits from INDIDriver.
+   * <code>false</code> otherwise.
+   */
+  private boolean isINDIDriver(Class c) {
+if (INDINotLoadableDriver.class.isAssignableFrom(c)) {
+  return false; 
+}
+    Class s = c.getSuperclass();
+
+    while (s != null) {
+      if (s == INDIDriver.class) {
+        return true;
+      }
+
+      s = s.getSuperclass();
+    }
+
+    return false;
+  }
+
+  /**
+   * Gets a list of Strings with the name of the classes in a JAR file.
+   *
+   * @param file The JAR file.
+   * @return A list with all the names of the classes in the JAR file.
+   * @throws INDIException if there is any problem accessing to the JAR file.
+   */
+  private ArrayList<String> getClassesInJAR(String file) throws INDIException {
+
+    ArrayList<String> list = new ArrayList<String>();
+
+    try {
+      JarInputStream jarFile = new JarInputStream(new FileInputStream(file));
+      JarEntry jarEntry = jarFile.getNextJarEntry();
+      while (jarEntry != null) {
+        if (jarEntry.getName().endsWith(".class")) {
+          String n = jarEntry.getName().replace("/", ".");
+
+          n = n.substring(0, n.length() - 6);
+
+          list.add(n);
+        }
+
+        jarEntry = jarFile.getNextJarEntry();
+      }
+    } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)